packages/cell-toolbar/src/celltoolbartracker.ts

```diff
@@ -0,0 +1,344 @@
+      return true;
+    }
+
+    if (cellType === 'markdown' && (activeCell as MarkdownCell).rendered) {
+      // Check for overlap in rendered markdown content
+      return this._markdownOverlapsToolbar(activeCell as MarkdownCell);
+    }
+
+    // Check for overlap in code content
+    return this._codeOverlapsToolbar(activeCell);
+  }
+
+  /**
+   * Check for overlap between rendered Markdown and the cell toolbar
+   *
+   * @param activeCell A rendered MarkdownCell
+   * @returns `true` if the first line of the output overlaps with the cell toolbar, `false` otherwise
+   */
+  private _markdownOverlapsToolbar(activeCell: MarkdownCell): boolean {
+    const markdownOutput = activeCell.inputArea; // Rendered markdown appears in the input area
+
+    // Get the rendered markdown as a widget.
+    const markdownOutputWidget = markdownOutput.renderedInput;
+    const markdownOutputElement = markdownOutputWidget.node;
+
+    const firstOutputElementChild = markdownOutputElement.firstElementChild as HTMLElement;
+    if (firstOutputElementChild === null) {
+      return false;
+    }
+
+    // Temporarily set the element's max width so that the bounding client rectangle only encompasses the content.
+    const oldMaxWidth = firstOutputElementChild.style.maxWidth;
+    firstOutputElementChild.style.maxWidth = 'max-content';
+
+    const lineRight = firstOutputElementChild.getBoundingClientRect().right;
+
+    // Reinstate the old max width.
+    firstOutputElementChild.style.maxWidth = oldMaxWidth;
+
+    const toolbarLeft = this._cellToolbarLeft(activeCell);
+
+    return toolbarLeft === null ? false : lineRight > toolbarLeft;
+  }
+
+  private _codeOverlapsToolbar(activeCell: Cell<ICellModel>): boolean {
+    const editorWidget = activeCell.editorWidget;
+    const editor = activeCell.editor;
+    if (editor.lineCount < 1) {
+      return false; // Nothing in the editor
+    }
+
+    const codeMirrorLines = editorWidget.node.getElementsByClassName(
+      'CodeMirror-line'
+    );
+    if (codeMirrorLines.length < 1) {
+      return false; // No lines present
+    }
+    const lineRight = codeMirrorLines[0].children[0] // First span under first pre
+      .getBoundingClientRect().right;
+
+    const toolbarLeft = this._cellToolbarLeft(activeCell);
+
+    return toolbarLeft === null ? false : lineRight > toolbarLeft;
+  }
+
+  private _cellEditorWidgetLeft(activeCell: Cell<ICellModel>): number {
+    return activeCell.editorWidget.node.getBoundingClientRect().left;
+  }
+
+  private _cellEditorWidgetRight(activeCell: Cell<ICellModel>): number {
+    return activeCell.editorWidget.node.getBoundingClientRect().right;
+  }
+
+  private _cellToolbarLeft(activeCell: Cell<ICellModel>): number | null {
+    const toolbarWidgets = this._findToolbarWidgets(activeCell);
+    if (toolbarWidgets.length < 1) {
+      return null;
+    }
+    const activeCellToolbar = toolbarWidgets[0].node;
+
+    return activeCellToolbar.getBoundingClientRect().left;
```

Comment:
I wonder if `getBoundingClientRect` could be cached (if line has not changed) to prevent potential performance degradation in a mechanism of long layout recalculation as discussed in https://github.com/jupyterlab/jupyterlab/issues/9757. I meant to open an issue but went directly to this PR to see if we caught anything in the performance benchmarks (likely it would not be picked up as it requires user to focus a cell which may not be sufficiently covered by the current scenarios), but I do not see reference values in https://github.com/jupyterlab/jupyterlab/pull/12028#issuecomment-1097874513; is this expected @fcollonval ?

Reply:
Good catch - the second run (on the PR verison) is missing the update of the reference snapshots. I will correct that. Unfortunately the [weekly benchmark](https://github.com/jupyterlab/benchmarks/issues/85#issuecomment-1100791897) will not provide much insight for this specific PR because the [PR turn off virtual rendering](https://github.com/jupyterlab/jupyterlab/pull/12258) was also merged the same week.